### PR TITLE
Add support for legacy table attribute cellpadding

### LINKED ISSUE #12609: Table cellpadding attribute
Although fairly legacy, it probably makes sense for servo to support the `cellpadding` attribute of `<table>`.

```
<html>
<head>
    <title>Table cellpadding attribute + padding CSS</title>
    <style type="text/css">
        table { background-color: red; }
        div { width:10px;height:10px; }
    </style>
</head>
<body>
    <table>
        <tr>
            <td style="padding:5px"><div></div></td>
        </tr>
    </table>
    <table cellpadding="5">
        <tr>
            <td><div></div></td>
        </tr>
    </table>
</body>
</html>
```

The above example renders as a single column of red in Firefox, Chrome & Edge.

https://html.spec.whatwg.org/multipage/rendering.html#tables-2 describes how cellpadding should be interpreted.

Related to #11821 in the sense that this prevents HN's top bar from being rendered correctly.

I'm hoping to give this a go by following how `cellspacing` was implemented in https://github.com/servo/servo/pull/4417 . Some of that has been moved around quite a bit since though.

### PR DESCRIPTION
**Pull request by adamncasey**
Handle the parent table element's `cellpadding` attribute value during table cell layout.

When `synthesize_presentational_hints_for_legacy_attributes` is called with an `HTMLTableCellElement`, look along a fixed path for the parent table element using `LayoutJS<Node>::parent_node_ref()`.

As touched on in #12609, another implementation worth evaluating would be for the `HTMLTableElementMethods::InsertRow` method to pass down the table's cellpadding value, and then for the `HTMLTableRowElementMethods::InsertCell` to pass it to each Cell. This would remove the need for the parent table lookup for every cell during layout, at the expense of extra memory per cell. This value needs to be mutable at the table level, as the property can be updated. Let me know if this method seems more appropriate. I haven't gone down this route so far because I wasn't clear on which of Rust's reference wrappers would be best suited for this.

Hacker News rendering improvement:
cellpadding

There are a few issues running the tests on Windows right now - no headless mode implemented, and seems a lot of the test expectations aren't right. I'm relying on the CI here.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #12609

- [X] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18161)

### CODE REVIEW

File: components/script/dom/element.rs [hunk 1]
```diff
@@ -746,6 +746,30 @@ impl LayoutElementHelpers for LayoutJS<Element> {
                 shared_lock,
                 PropertyDeclaration::BorderRightWidth(width_value)));
         }
+
+        let cellpadding = if let Some(this) = self.downcast::<HTMLTableCellElement>() {
```
Comment (emilio) [inline components/script/dom/element.rs]: We should seriously make this function virtual, it's really branchy already...

Of course not your fault.

File: components/script/dom/element.rs [hunk 2]
```diff
@@ -746,6 +746,30 @@ impl LayoutElementHelpers for LayoutJS<Element> {
...
+            this.get_inherited_cellpadding()
+        } else {
+            None
+        };
+
+        if let Some(cellpadding) = cellpadding {
```
Comment (emilio) [inline components/script/dom/element.rs]: This is in no way handling dynamic changes of any kind, right?

File: components/script/dom/htmltablecellelement.rs [hunk 3]
```diff
@@ -123,6 +128,20 @@ impl HTMLTableCellElementLayoutHelpers for LayoutJS<HTMLTableCellElement> {
                 .unwrap_or(LengthOrPercentageOrAuto::Auto)
         }
     }
+
+    fn get_inherited_cellpadding(&self) -> Option<u32>
+    {
+        unsafe {
+            // Follow chain: Cell -> Row -> Section -> Table
+            self.upcast::<Node>().parent_node_ref()
```
Comment (emilio) [inline components/script/dom/htmltablecellelement.rs]: Hmm... This works on a well-formed table, but what should happen if, for example, the row is anonymous? What do other browsers / the spec do?

File: components/script/dom/htmltablecellelement.rs [hunk 4]
```diff
@@ -123,6 +128,20 @@ impl HTMLTableCellElementLayoutHelpers for LayoutJS<HTMLTableCellElement> {
                 .unwrap_or(LengthOrPercentageOrAuto::Auto)
         }
     }
+
+    fn get_inherited_cellpadding(&self) -> Option<u32>
+    {
```
Comment (emilio) [inline components/script/dom/htmltablecellelement.rs]: nit: brace to the previous line.

File: components/script/dom/element.rs (hunk 2 above)
Comment (emilio) [inline components/script/dom/element.rs]: I think to do this you need to add this to add another case in `Document::element_attr_will_change`... But again it should be a virtual function.

**Comment by emilio** [comment]
Also, to make this green on CI you need to run `./mach update-manifest`. Thanks a lot for the patch!

File: components/script/dom/htmltablecellelement.rs (hunk 3 above)
Comment (adamncasey) [inline components/script/dom/htmltablecellelement.rs]: I believe tables are always constructed with this structure in Servo. If the input document's table isn't well formed, these elements are still inserted into the tree. This patch is OK in the anonymous row example, and where there are nested table rows.

Looking into other browsers, Firefox takes a similar route, with an additional case for where there is no HTMLTableSectionElement equivalent.
https://dxr.mozilla.org/mozilla-central/source/dom/html/HTMLTableCellElement.cpp#119
https://dxr.mozilla.org/mozilla-central/source/dom/html/HTMLTableCellElement.cpp#47
I'm not sure we need that here - I think there is always a tbody between the table and row.
https://github.com/servo/servo/blob/81a382cb00134450f3723caf682bb5ca8cca317f/components/script/dom/htmltableelement.rs#L306

Chrome searches up the tree until it finds a table element. This approach does need less code.
https://cs.chromium.org/chromium/src/third_party/WebKit/Source/core/html/HTMLTablePartElement.cpp?l=110&rcl=6ce68a3fb3e459bb4e9676a508a9260d4e67bd10
The short loop might be faster on average. I'm not sure how I'd feel about the possibility to search all the way up the tree though.

The spec isn't particularly helpful here https://www.w3.org/TR/html4/struct/tables.html#adef-cellpadding

**Comment by adamncasey** [comment]
Thanks for the review. I've had a go at fixing the manifest. Will take a longer look at the other points

File: components/script/dom/htmltablecellelement.rs (hunk 3 above)
Comment (jdm) [inline components/script/dom/htmltablecellelement.rs]: Note: https://html.spec.whatwg.org/#tables-2:attr-table-cellpadding is the spec we're using, not the w3c's specification.

**Comment by adamncasey** [comment]
7b91f4c adds support for dynamic changes to cellpadding. When any cellpadding attribute is changed (this is not filtered on `HTMLTableElement`), all children are restyled. In case anyone other than me finds this non-obvious: a re-cascade isn't enough to trigger the call to `synthesize_presentational_hints_for_legacy_attributes` on the cell elements.

This does not address @emilio's suggestion to make two functions touched in this patch (or parts of them) virtual.

**Comment by emilio** [comment]
Do you plan to do that in this PR? Or should I file a followup issue?

Yeah, cellpading is so annoying, you're right that recascading isn't enough, we also have this special-case on Gecko: http://searchfox.org/mozilla-central/rev/2aa0806c598ec433e431728f5ddd3a6847c1a417/layout/base/ServoRestyleManager.cpp#1374

Let me know if you think it's ready for review, happy to take a look on monday if so :)

Thanks a lot!

File: components/script/dom/htmltablecellelement.rs [hunk 5]
```diff
@@ -123,6 +128,19 @@ impl HTMLTableCellElementLayoutHelpers for LayoutJS<HTMLTableCellElement> {
                 .unwrap_or(LengthOrPercentageOrAuto::Auto)
         }
     }
+
+    fn get_inherited_cellpadding(&self) -> Option<u32> {
+        unsafe {
+            // Follow chain: Cell -> Row -> Section -> Table
+            self.upcast::<Node>().parent_node_ref()
```
Comment (emilio) [inline components/script/dom/htmltablecellelement.rs]: I'm still not sure whether this is the right thing to do... This works when it's a perfect table, but Firefox does handle a few more cases we should probably also handle:

http://searchfox.org/mozilla-central/rev/2aa0806c598ec433e431728f5ddd3a6847c1a417/dom/html/HTMLTableCellElement.cpp#46

**Comment by nox** [comment]
@emilio What's blocking this?

**Comment by emilio** [comment]
@nox mostly fixing https://github.com/servo/servo/pull/18161#discussion_r136723627, I'd think.

**Comment by Loirooriol** [comment]
Closing in favor of #31201

**Pull request closed without merging**